src/saga-dense.h

```diff
@@ -122,72 +124,97 @@ Saga(Penalty&               penalty,
      unsigned&              n_iter,
      std::vector<unsigned>& return_codes,
      std::vector<double>&   losses,
-     const bool             debug) noexcept
+     const bool             debug,
+     const bool             cyclic,
+     const unsigned         B) noexcept
 {
   using namespace std;
-
+    
   double wscale = 1.0;
-
+    
   double wscale_update = 1.0 - alpha*gamma;
-
+    
   penalty.setParameters(gamma, alpha, beta);
-
+    
   // Gradient vector and change in gradient vector
-  Eigen::ArrayXd g        = Eigen::ArrayXd::Zero(n_classes);
-  Eigen::ArrayXd g_change = Eigen::ArrayXd::Zero(n_classes);
-
-  Eigen::ArrayXd linear_predictor(n_classes);
-
+  Eigen::ArrayXXd g                = Eigen::ArrayXXd::Zero(n_classes, B);
+  Eigen::ArrayXXd g_change         = Eigen::ArrayXXd::Zero(n_classes, B);
+    
+  Eigen::ArrayXXd linear_predictor = Eigen::ArrayXXd::Zero(n_classes, B);
+    
+  // change for weight and average gradient 
+  Eigen::ArrayXXd step             = Eigen::ArrayXXd::Zero(n_classes, n_features);
+    
   // Setup functor for checking convergence
   ConvergenceCheck convergence_check{w, tol};
-
+    
+  // Setup selected sample matrix
+  Eigen::MatrixXd subx = Eigen::MatrixXd::Zero(n_features, B);
+    
+  // epoch size
+  const unsigned epoch = floor(n_samples/B);  
+    
+  // Setup index generator
+  Eigen::ArrayXXi index = Eigen::ArrayXXi::Zero(B, epoch);
+  Eigen::ArrayXi  s_ind = Eigen::ArrayXi::Zero(B);
+    
   // Outer loop
   unsigned it_outer = 0;
   bool converged = false;
   do {
+      
+    // Pull samples
+    index = Index(n_samples, B, cyclic);
```

Comment:
sb `Eigen::Index`

Reply:
Sorry about this, I have changed its name to "Ind" now.